aten/src/ATen/native/cuda/Loss.cu

```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+    *output = -cur_weight * input[t];
+    *total_weight = cur_weight;
+    if (size_average && *total_weight > 0) {
+      *output /= *total_weight;
+    }
+  }
+}
+
+template <typename scalar_t, typename accscalar_t>
+__global__ void nll_loss_forward_reduce_cuda_kernel_2d(
+    scalar_t* output,
+    scalar_t* total_weight,
+    scalar_t* input,
+    int64_t* target,
+    scalar_t* weights,
+    bool size_average,
+    int nframe,
+    int ndim,
+    int n_classes,
+    int64_t ignore_index) {
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  __shared__ accscalar_t sh_inputs[NLL_LOSS_THREADS],
+      acc_weight[NLL_LOSS_THREADS];
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  int i, t;
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  scalar_t cur_weight;
+
+  sh_inputs[threadIdx.x] = static_cast<accscalar_t>(0);
+  acc_weight[threadIdx.x] = static_cast<accscalar_t>(0);
+  for (i = threadIdx.x; i < nframe; i += NLL_LOSS_THREADS) {
+    t = target[i];
+    if (t != static_cast<int>(ignore_index)) {
+      CUDA_KERNEL_ASSERT(t >= 0 && t < n_classes);
+      cur_weight = weights != nullptr ? weights[t] : static_cast<scalar_t>(1);
+      sh_inputs[threadIdx.x] -= input[i * ndim + t] * cur_weight;
+      acc_weight[threadIdx.x] += cur_weight;
+    }
+  }
+
+  __syncthreads();
+
+  if (threadIdx.x == 0) {
+    *output = *total_weight = static_cast<scalar_t>(0);
+    accscalar_t output_acc = 0;
+    accscalar_t total_weight_acc = 0;
+    for (i = 0; i < NLL_LOSS_THREADS; ++i) {
+      output_acc += sh_inputs[i];
+      total_weight_acc += acc_weight[i];
+    }
+    *total_weight = static_cast<scalar_t>(total_weight_acc);
+    *output = static_cast<scalar_t>(output_acc);
+    if (size_average) {
+      if (nframe == 0) {
+        // Mean reduction on empty tensors produces NaN
+        *output = std::numeric_limits<double>::quiet_NaN();
+      }
+      if (*total_weight != 0) {
+        *output = static_cast<scalar_t>(output_acc / total_weight_acc);
+      }
+    }
+  }
+}
+
+} // namespace
+
+void nll_loss_forward_out_cuda_template(
+    Tensor& output,
+    Tensor& total_weight,
+    const Tensor& input,
+    const Tensor& target,
+    const c10::optional<Tensor>& weight_opt,
+    int64_t reduction,
+    int64_t ignore_index) {
+  // See [Note: hacky wrapper removal for optional tensor]
+  c10::MaybeOwned<Tensor> weight_maybe_owned =
+      at::borrow_from_optional_tensor(weight_opt);
+  const Tensor& weight = *weight_maybe_owned;
+
+  TORCH_CHECK(target.dim() <= 1, "multi-target not supported");
+
+  int64_t n_classes = input.size(-1);
+  int64_t n_dims = input.dim();
+
+  TensorArg output_arg{output, "output", 1};
+  TensorArg total_weight_arg{total_weight, "total_weight", 2};
+  TensorArg input_arg{input, "input", 3};
+  TensorArg target_arg{target, "target", 4};
+
+  if (weight.defined()) {
+    TensorArg weight_arg{weight, "weight", 5};
+    checkAllSameGPU(
```

Comment:
the device placement of all the args is automatically checked in codegen, you don't have to do it here (that also means you don't have to create TensorArgs